test/redis/windowed_cache_impl_test.go

```diff
@@ -0,0 +1,340 @@
+			{Code: pb.RateLimitResponse_OK, CurrentLimit: limits[0].Limit, LimitRemaining: 2, DurationUntilReset: &duration.Duration{Seconds: 13 * 4 * 60}}},
+		cache.DoLimit(nil, request, limits))
+	assert.Equal(uint64(2), limits[0].Stats.TotalHits.Value())
+	assert.Equal(uint64(0), limits[0].Stats.OverLimit.Value())
+	assert.Equal(uint64(0), limits[0].Stats.OverLimitWithLocalCache.Value())
+	assert.Equal(uint64(1), limits[0].Stats.NearLimit.Value())
+
+	// Check the local cache stats.
+	testLocalCacheStats(localCacheStats, statsStore, sink, 0, 2, 2, 0, 0)
+
+	// Test Over limit stats
+	timeSource.EXPECT().UnixNanoNow().Return(int64(60 * 4 * 60 * 1e9)).MaxTimes(1)
+	client.EXPECT().PipeAppend(gomock.Any(), gomock.Any(), "SETNX", "domain_key_value_0", int64(0)).DoAndReturn(pipeAppend)
+	client.EXPECT().PipeAppend(gomock.Any(), gomock.Any(), "EXPIRE", "domain_key_value_0", int64(60*60)).DoAndReturn(pipeAppend)
+	client.EXPECT().PipeAppend(gomock.Any(), gomock.Any(), "GET", "domain_key_value_0").SetArg(1, int64(75*4*60*1e9)).DoAndReturn(pipeAppend)
+	client.EXPECT().PipeDo(gomock.Any()).Return(nil)
+
+	request = common.NewRateLimitRequest("domain", [][][2]string{{{"key", "value"}}}, 1)
+
+	limits = []*config.RateLimit{
+		config.NewRateLimit(15, pb.RateLimitResponse_RateLimit_HOUR, "key_value", statsStore)}
+
+	assert.Equal(
+		[]*pb.RateLimitResponse_DescriptorStatus{
+			{Code: pb.RateLimitResponse_OVER_LIMIT, CurrentLimit: limits[0].Limit, LimitRemaining: 0, DurationUntilReset: &duration.Duration{Seconds: 15 * 4 * 60}}},
+		cache.DoLimit(nil, request, limits))
+	assert.Equal(uint64(3), limits[0].Stats.TotalHits.Value())
+	assert.Equal(uint64(1), limits[0].Stats.OverLimit.Value())
+	assert.Equal(uint64(0), limits[0].Stats.OverLimitWithLocalCache.Value())
+	assert.Equal(uint64(1), limits[0].Stats.NearLimit.Value())
+
+	// Check the local cache stats.
+	testLocalCacheStats(localCacheStats, statsStore, sink, 0, 2, 3, 0, 1)
+
+	// Test Over limit stats with local cache
+	timeSource.EXPECT().UnixNanoNow().Return(int64(60 * 4 * 60 * 1e9)).MaxTimes(1)
+
+	request = common.NewRateLimitRequest("domain", [][][2]string{{{"key", "value"}}}, 1)
+
+	limits = []*config.RateLimit{
+		config.NewRateLimit(15, pb.RateLimitResponse_RateLimit_HOUR, "key_value", statsStore)}
+
+	assert.Equal(
+		[]*pb.RateLimitResponse_DescriptorStatus{
+			{Code: pb.RateLimitResponse_OVER_LIMIT, CurrentLimit: limits[0].Limit, LimitRemaining: 0, DurationUntilReset: &duration.Duration{Seconds: 15 * 4 * 60}}},
+		cache.DoLimit(nil, request, limits))
+	assert.Equal(uint64(4), limits[0].Stats.TotalHits.Value())
+	assert.Equal(uint64(2), limits[0].Stats.OverLimit.Value())
+	assert.Equal(uint64(1), limits[0].Stats.OverLimitWithLocalCache.Value())
+	assert.Equal(uint64(1), limits[0].Stats.NearLimit.Value())
+
+	// Check the local cache stats.
+	testLocalCacheStats(localCacheStats, statsStore, sink, 1, 3, 4, 0, 1)
+}
+
+func TestRedisWindowedWithJitter(t *testing.T) {
+	assert := assert.New(t)
```

Comment:
can the code which sets up all the mocks and creates cache impl object be pulled from each test into some sort of SetUp method and be called in each test?